### ISSUE TITLE
Code for filtering ship groups increases Pyfa boot time dramatically

### ISSUE THREAD

**Issue by Ebag333** [pyfa-org/Pyfa#819]
So we have some code that lets you click a button to hide ship groups you have no fits in.

This code is _bad_.

For my primary database, this code adds about 5s to the load time (and I have a ship in nearly every market group).

For an empty database...well....this.  (I clicked to load a ship the instant I could, just to get the timestamp.)
```
2016-11-16 21:06:26,394 root                     INFO     Starting pyfa
2016-11-16 21:07:11,579 service.fit              DEBUG    ==========recalc==========
```

Yes, it took nearly a minute to open Pyfa with an empty database.

What this code does is loop through ever market group, then every ship in the market group.  If it finds a ship you have fits for in that market group, it skips to the next one.

I'd like to burn this code with fire.  This feature isn't particularly useful, and as CCP continues to add new ships/groups it will continue to slow down launching Pyfa.  What say you, @blitzmann ?

**Comment by blitzmann** [comment]
> Yes, it took nearly a minute to open Pyfa with an empty database.

There is something seriously wrong with your machine then. It's very odd how you always have performance issues. Are you debugging, or just starting pyfa from terminal?

I have a somewhat empty database (like 10 fits) and mine starts in 6 seconds. 

'''
2016-11-17 16:41:44,680 root                     INFO     Starting pyfa
2016-11-17 16:41:50,647 service.fit              DEBUG    ==========recalc==========
'''

I don't remember exactly how the code works, but it might be possible to get more performance if we run a direct sql query rather than using ORM, and then apply that, although there is the hiccup that the data is separate. I don't want to touch this right now, but will leave this open for further investigation. This was a feature that was asked for by many people, so I'd rather not get rid of it outright.

**Comment by blitzmann** [comment]
For example:

```
SELECT g.groupID, g.groupName, COUNT(*) 
FROM fits f
JOIN eve.invTypes t ON f.shipID = t.typeID
JOIN eve.invGroups g ON t.groupID = g.groupID
GROUP BY g.groupID, g.groupName
```

Gets us the number of fits per group, fast and easy, instead of looping over the shit. So we could run this query, and then assign the number of fits to the marketgroups. 

To do this, we would need to attach the eve database to the user database, but we can detach it immediately afterwards. (or leave it attached? Dunno if that can happen). I don't know the implications of this, but this is a route we could potentially take.

But yeah, even if that shaves off 5 seconds from your launch... what the hell is causing it to take a minute to launch?

**Comment by Ebag333** [comment]
>There is something seriously wrong with your machine then. It's very odd how you always have performance issues. 

You always blame my machine. :)

There's not.  It's a fresh install of Win10, very vanilla.  The hardware is less than a year old, so it's a 1 gen old i5 processor with 8 GB of RAM and a decently fast SSD.

My "performance issues" are consistent as well. The RAH times that I calculated didn't change after the recent fresh install.  Running in a virtual env vs against the installed Python env directly doesn't change anything either.  Obviously not running with debugging turned on and not running in debug mode speeds things up, but the difference there is consistent as well.

>Are you debugging, or just starting pyfa from terminal?

As I mentioned, same code base running my normal DB (which is over a meg now!) this step takes 4-5 seconds to run.  I've done some profiling on it because the startup delay seriously annoys me, since I'm constantly restarting Pyfa.

I am running in debug mode, from pyCharm, so it's definitely going to be slower than "normal", but this is still slow enough to be noticeable otherwise.

This isn't a one off thing, either,  I launched it many times and saw the same issue repeatedly.

I know that this feature was asked for, but it's the single largest point to slow down in the boot up process (aside from running the fitting recalc engine twice against nothing, don't ask me why...).  I did some fairly in depth profiling, which is how I figured this out in the first place.

The way the code is written, every time there's a new market group, and (potentially) every time there is a new ship, this will get slower (because,  more loops).

I poked around a bit at speeding up the processing, but I don't see any quick and easy ways to do that.  It's just the nature what we're doing.

Psuedo code, it's essentially doing this:
```
for Group in MarketGroups{
     Get Ships in Group
     for Ship in Ships{
          Get Fits for Ship
          If Fits>0
               MarkGroupAsHavingFits()
               Break
     }
}
```

>Gets us the number of fits per group, fast and easy, instead of looping over the shit. So we could run this query, and then assign the number of fits to the marketgroups.

We can probably just implement this as a SQLAlchemy query.

>To do this, we would need to attach the eve database to the user database, but we can detach it immediately afterwards. (or leave it attached? Dunno if that can happen). I don't know the implications of this, but this is a route we could potentially take.

I actually wanted to investigate handling the queries to query both DBs at the same time.  That would help a lot with some fairly painful work arounds we have to query each DB separately, then connect the two.  SQLite can handle it fine, as can MS SQL and MySQL, I just don't know off hand if SQLAlchemy supports it.

>But yeah, even if that shaves off 5 seconds from your launch... what the hell is causing it to take a minute to launch?

It's just the loop within loop within loop that we have.  If there are no ships within the user DB, it has to literally loop through every ship in every market group.

This is why as CCP adds more groups and more ships this gets slower and slower...and....slower....and...........

**Comment by blitzmann** [comment]
>You always blame my machine. :)

Sorry, I don't mean to place blame. But it's just so odd how you are always getting slow performance. And I agree with you, it is consistent. But having a >1 minute start up is not normal and not something I've ever experienced on my 4-year old build, or even on my slow as balls OS X VM. I don't know the reason, but I strongly suspect it's localized - if pyfa's startup time was >1 minutes for others, I feel like people would speak up about it.

> I am running in debug mode, from pyCharm, so it's definitely going to be slower than "normal", but this is still slow enough to be noticeable otherwise.

This is probably part of the problem. Running the debugger in pycharm causes mine to slow to a crawl when first launching. Still, it only takes about 10-15 seconds depending on how many fits I have open on startup. Then again, that's on an almost empty database, and I don't have a large one to test with right now.

Can you do me a favor? It's kind of a hassle, but could you log the startup times under the following conditions (always having `-d` flag of course):

With current code
    - Without debugger, fresh database (or have 1 fit that is open so you capture the first recalc message)
    - With debugger, fresh database
    - Without debugger, large database
    - With debugger, large database

And then do the same with the code in question commented out. 

Please post results here. Again, I'm not wanting to remove features and functionality. But I can't reproduce the problem you're having (also, can you point out where this code is called so that I can also investigate without spending 15 minutes hunting it down :P). I'm not saying it's not a problem, that code can be more efficient of course, but I'm just baffled by why your installation always has performance issues?

> We can probably just implement this as a SQLAlchemy query.

Yeah, my thought was to just make a direct query in one of the `queries.py` files. This will be called as needed, and is just informational - doesn't really need to the ORM structure. It can basically act as a lookup table, return a dictionary of `margetGroupID: numOfFits`

>I actually wanted to investigate handling the queries to query both DBs at the same time. That would help a lot with some fairly painful work arounds we have to query each DB separately, then connect the two.

I think it could be beneficial, but this would be a much, much large change to the very internals of pyfa. We're talking taking a lot of what's in `eos` and completely refactoring it. pyfa is in late life, and it's not something I'm particularly interested in doing, but if you want to look into it and find a few neat tricks we can do, I can get behind that. But right now `eos` is under the assumption of two separate database. ;_;

**Comment by Ebag333** [comment]
>I feel like people would speak up about it.

Probably because Pyfa was slow to start with, and just got slower over time (as most programs tend to do).

Slightly off topic, but I worked at a place with 3500 users, and there was an ongoing issue where they'd occasionally fail to login, and if it did login would take a good 5-10 minutes.  We had 6 complaints after months of this occurring.  People just don't bring this sort of thing up, especially when it just slowly gets worse over time.

Anyway, here's the relevant bits of code that I found took significant chunks of time when loading Pyfa.

https://github.com/pyfa-org/Pyfa/blob/master/service/market.py#L555

https://github.com/pyfa-org/Pyfa/blob/master/service/fit.py#L149

This is pretty spaghetti code where you have chains of functions that each do one thing, then call another function.

I'll see about testing this further,  I do want to get to the bottom of it, but getting the capsim and patch done is a higher priority at the moment.
